functional-tests/appwrapper_reconciler_test.go

```diff
@@ -0,0 +1,228 @@
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	log "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	ctx    context.Context
+	cancel context.CancelFunc
+)
+
+func logger() {
+	// Get log file path from environment variable or use a default path
+	logFilePath := os.Getenv("LOG_FILE_PATH")
+	if logFilePath == "" {
+		logFilePath = "./functional-test-logfile.log"
+	}
+	// Create a log file
+	logFile, err := os.Create(logFilePath)
+	if err != nil {
+		log.Log.Error(err, "Error creating log file: %v", err)
+	}
+	// Configure zap logger to write to the log file
+	logger := zap.New(zap.WriteTo(logFile), zap.UseDevMode(true))
+
+	// Set the logger for controller-runtime
+	ctrl.SetLogger(logger)
+
+	// This line prevents controller-runtime from complaining about log.SetLogger never being called
+	log.SetLogger(logger)
+}
+
+func startEnvTest(t *testing.T) *rest.Config {
+	// to redirect all functional test related logs to separate logfile ~ default (in functional-tests directory), can be changed using environment variable LOG_FILE_PATH=/path_to_logfile
+	logger()
+
+	//specify testEnv configuration
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "project-codeflare", "multi-cluster-app-dispatcher@v1.39.0", "config", "crd", "bases"),
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "openshift", "api@v0.0.0-20230213134911-7ba313770556", "machine", "v1beta1"),
+		},
+		ErrorIfCRDPathMissing: true,
+	}
+	test := WithConfig(t, testEnv.Config)
+	cfg, err := testEnv.Start()
+	test.Expect(err).NotTo(HaveOccurred())
+	test.Expect(cfg).NotTo(BeNil())
+
+	t.Cleanup(func() {
+		teardownTestEnv(test, testEnv)
+	})
+
+	return cfg
+}
+
+func startInstascaleController(test Test, cfg *rest.Config) {
+	ctx, cancel = context.WithCancel(context.TODO())
+
+	// Add custom resource schemes to the global scheme.
+	err := mcadv1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	err = machinev1beta1.AddToScheme(scheme.Scheme)
+	test.Expect(err).NotTo(HaveOccurred())
+
+	// Create a controller manager for managing controllers.
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme.Scheme,
+	})
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Create an instance of the AppWrapperReconciler with configuration
+	instaScaleController := &controllers.AppWrapperReconciler{
+		Client: k8sManager.GetClient(),
+		Scheme: k8sManager.GetScheme(),
+		Config: config.InstaScaleConfiguration{
+			MachineSetsStrategy: "reuse",
+			MaxScaleoutAllowed:  5,
+		},
+	}
+	// Set up the AppWrapperReconciler with the manager.
+	err = instaScaleController.SetupWithManager(context.Background(), k8sManager)
+	test.Expect(err).ToNot(HaveOccurred())
+
+	// Start the controller manager in a goroutine.
+	go func() {
+		defer GinkgoRecover()
```

Comment:
AFAIK this is not needed, can you remove the GinkgoRecover and relevant entries in go.mod and go.sum?